src/lib/swapper/swappers/ThorchainSwapper/utils/makeMemoWithShortenedFinalAssetAmount.ts

```diff
@@ -0,0 +1,91 @@
+import assert from 'assert'
+import { toBaseUnit } from 'lib/math'
+import { assertAndProcessMemo } from 'lib/utils/thorchain/memo'
+
+import { AGGREGATOR_EXPONENT_LENGTH, THORCHAIN_PARSER_MAXIMUM_PRECISION } from '../constants'
+import { MEMO_PART_DELIMITER } from './constants'
+
+// Allows us to convert a THORChain parser value back to its base unit
+// for testing purposes
+export const thorchainParserToBaseUnit = (thorchainParserValue: string) => {
+  if (thorchainParserValue.length < 3) {
+    throw new Error('Number too short to contain an exponent.')
+  }
+
+  const exponent = parseInt(thorchainParserValue.slice(-2), 10)
+
+  const baseNumber = thorchainParserValue.slice(0, -2)
+
+  const fullNumber = toBaseUnit(baseNumber, exponent === 1 ? 0 : exponent)
+
+  return fullNumber
+}
+
+export const makeMemoWithShortenedFinalAssetAmount = ({
+  maxMemoSize,
+  memoWithoutMinAmountOut,
+  finalAssetLimitWithManualSlippage,
+}: {
+  maxMemoSize: number
+  memoWithoutMinAmountOut: string
+  finalAssetLimitWithManualSlippage: string
+}) => {
+  const HYPOTHETICAL_EXPONENT = '01'
+
+  const memoArrayWithoutMinAmountOut = memoWithoutMinAmountOut.split(MEMO_PART_DELIMITER)
+
+  // We need to construct the memo with the minAmountOut at the end so we can calculate his bytes size
+  const unshortenedMemo = `${memoWithoutMinAmountOut}:${finalAssetLimitWithManualSlippage}${HYPOTHETICAL_EXPONENT}`
+
+  const memoSizeLeft = maxMemoSize - unshortenedMemo.length
+
+  let excessBytesToTrim = 0
+
+  if (memoSizeLeft < 0) {
+    excessBytesToTrim = Math.abs(memoSizeLeft)
+  }
+
+  const shortenedMinAmountOutLength =
+    finalAssetLimitWithManualSlippage.length + AGGREGATOR_EXPONENT_LENGTH - excessBytesToTrim
+
+  // THORChain parser uses big.ParseFloat with `0` as a precision, resulting in a precision of 18 + the integer part, so we can't more than 19 characters
+  // https://gitlab.com/thorchain/thornode/-/blob/v1.131.0/x/thorchain/memo/memo_parser.go#L190
+  if (shortenedMinAmountOutLength > THORCHAIN_PARSER_MAXIMUM_PRECISION + 1) {
+    excessBytesToTrim += shortenedMinAmountOutLength - (THORCHAIN_PARSER_MAXIMUM_PRECISION + 1)
+  }
+
+  const shortenedAmountOut = finalAssetLimitWithManualSlippage.substring(
+    0,
+    finalAssetLimitWithManualSlippage.length - excessBytesToTrim,
+  )
+
+  // Paranoia check - we should never have a 0 amount out
+  assert(shortenedAmountOut !== '0', 'expected final amount limit to be different than 0')
```

Comment:
Do we also want to paranoia check that we don't have `001` either, i.e `0` + `01` exponent?

Reply:
We already check for this under the hood earlier as we check the amount without the exponent